### PR TITLE
cmd, lib/game: Parallel universe

### PR DESCRIPTION
**Pull request by deadprogram**
This PR adds a `ParallelUniverse` type to `lib/game` that uses channels to connect multiple parallel universes. It also adds the ability to run them in the `vita` command line tool.

### CODE REVIEW

**Comment by deadprogram** [comment]
Tests fixed and rebased.

File: lib/game/parallel.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	for _, neighborRow := range []int32{r - 1, r, r + 1} {
+		for _, neighborColumn := range []int32{c - 1, c, c + 1} {
+			if neighborRow == r && neighborColumn == c {
+				// Skip checking the cell itself
+				continue
+			}
+
+			switch {
+			case neighborRow < 0 && neighborColumn < 0:
+				// check the universe above and to the left?
```
Comment (acifani) [inline lib/game/parallel.go:60]: we either have to also connect diagonal universes or expose a function to check for these cases

File: lib/game/parallel.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
+package game
+
+import (
+	"sync"
+)
+
+type NeighborData struct {
+	Generation uint32
+	Data       []uint8
```
Comment (acifani) [inline lib/game/parallel.go:9]: nitpick: does not need to be now, but I'd give anything to avoid the dreaded `.Data.Data`

File: lib/game/parallel.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+				// check the universe below and to the left?
+			case neighborRow >= int32(p.height) && neighborColumn >= int32(p.width):
+				// check the universe below and to the right?
+			case neighborRow < 0:
+				// check the universe above
+				if p.TopNeighbor == nil {
+					continue
+				}
+
+				neighborIdx := p.GetIndex(p.height-1, uint32(neighborColumn))
```
Comment (acifani) [inline lib/game/parallel.go:73]: Here we are implying the universes are always the same size right?

File: lib/game/parallel.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+// ParallelNeighbors returns the number of alive neighbors for a given cell.
+// It uses the Moore neighborhood, which includes the eight cells surrounding
+// the given cell, but also extended to adjoining neighbor Universes.
+func (p *ParallelUniverse) ParallelNeighbors(row, column uint32) uint8 {
+	count := uint8(0)
+	r, c := int32(row), int32(column)
+	for _, neighborRow := range []int32{r - 1, r, r + 1} {
+		for _, neighborColumn := range []int32{c - 1, c, c + 1} {
+			if neighborRow == r && neighborColumn == c {
```
Comment (acifani) [inline lib/game/parallel.go:53]: maybe we can add this as the first case in the switch?

File: lib/game/parallel.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+			case neighborRow < 0:
+				// check the universe above
+				if p.TopNeighbor == nil {
+					continue
+				}
+
+				neighborIdx := p.GetIndex(p.height-1, uint32(neighborColumn))
+				if p.TopNeighbor.Data.Data[neighborIdx] != Dead {
+					count++
+				}
```
Comment (acifani) [inline lib/game/parallel.go:76]: question: would a `break `help speed things up here? not too familiar with that

File: cmd/vita/main.go [hunk 6]
```diff
@@ -25,3 +41,61 @@ func main() {
...
+			fmt.Println()
+		}
+
+		var wg sync.WaitGroup
+		for _, u := range multi[:len(multi)-1] {
+			callMultiTick(&wg, u)
+		}
+		wg.Wait()
+
+		multi[len(multi)-1].MultiTick()
```
Comment (acifani) [inline cmd/vita/main.go:59]: question: why is this needed?

File: lib/game/universe.go [hunk 7]
```diff
@@ -10,11 +10,12 @@ const (
...
-	width    uint32
-	cells    []uint8
-	newCells []uint8
-	stable   bool
+	height     uint32
+	width      uint32
+	cells      []uint8
+	newCells   []uint8
+	stable     bool
+	Generation uint32
```
Comment (acifani) [inline lib/game/universe.go:18]: I don't think we are currently using this, right? Is there any plan to or can we remove it?

**Review by acifani — APPROVED** [review]
great stuff! I love that we can implement different transport methods by overriding a few functions in the future

File: lib/game/parallel.go (hunk 1 above)
Comment (deadprogram) [reply lib/game/parallel.go:60]: It is a check for only 1 pixel which requires connecting to a whole `Universe` :crying_cat_face: 

Not sure what to do about it at the moment, hence why the comment. Perhaps ignore for now?

File: lib/game/parallel.go (hunk 3 above)
Comment (deadprogram) [reply lib/game/parallel.go:73]: That is correct. Might want to enforce that in the future.

File: lib/game/parallel.go (hunk 4 above)
Comment (deadprogram) [reply lib/game/parallel.go:53]: Good idea!

File: lib/game/parallel.go (hunk 5 above)
Comment (deadprogram) [reply lib/game/parallel.go:76]: Not needed in the Go `switch` statement.

File: cmd/vita/main.go (hunk 6 above)
Comment (deadprogram) [reply cmd/vita/main.go:59]: It isn't actually. I just consolidated that into the main loop.

File: lib/game/universe.go (hunk 7 above)
Comment (deadprogram) [reply lib/game/universe.go:18]: I am planning to use it for an upcoming feature, which is why I added it.

File: lib/game/parallel.go (hunk 2 above)
Comment (deadprogram) [reply lib/game/parallel.go]: Yeah, I also dislike that. Changed to `Cells`

**Comment by deadprogram** [comment]
Now merging, thanks for the feedback @acifani
